File: app/src/main/java/com/example/harapp/HARModel.java
package com.example.harapp;

import android.content.Context;
import org.pytorch.IValue;
import org.pytorch.LiteModuleLoader;
import org.pytorch.Module;
import org.pytorch.Tensor;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

public class HARModel {
    private Module model;
    private static final int SEQUENCE_LENGTH = 128;
    private static final int NUM_FEATURES = 9;
    private static final int NUM_CLASSES = 6;

    // Activity labels matching UCI HAR dataset
    private static final String[] ACTIVITY_LABELS = {
            "WALKING",
            "WALKING_UPSTAIRS",
            "WALKING_DOWNSTAIRS",
            "SITTING",
            "STANDING",
            "LAYING"
    };

    public HARModel(Context context) throws IOException {
        model = LiteModuleLoader.load(assetFilePath(context, "har_model.ptl"));
    }

    /**
     * Predict activity from sensor data
     * @param sensorData float array of shape [128, 9]
     * @return PredictionResult containing activity and confidence
     */
    public PredictionResult predict(float[][] sensorData) {
        // Validate input
        if (sensorData.length != SEQUENCE_LENGTH || sensorData[0].length != NUM_FEATURES) {
            throw new IllegalArgumentException(
                    "Expected input shape [128, 9], got [" +
                            sensorData.length + ", " + sensorData[0].length + "]"
            );
        }

        // Flatten to 1D array
        float[] inputArray = new float[SEQUENCE_LENGTH * NUM_FEATURES];
        for (int i = 0; i < SEQUENCE_LENGTH; i++) {
            System.arraycopy(sensorData[i], 0, inputArray, i * NUM_FEATURES, NUM_FEATURES);
        }

        // Create tensor
        Tensor inputTensor = Tensor.fromBlob(
                inputArray,
                new long[]{1, SEQUENCE_LENGTH, NUM_FEATURES}
        );

        // Run inference
        Tensor outputTensor = model.forward(IValue.from(inputTensor)).toTensor();
        float[] scores = outputTensor.getDataAsFloatArray();

        // Apply softmax to get probabilities
        float[] probabilities = softmax(scores);

        // Find max probability
        int maxIndex = 0;
        float maxProb = probabilities[0];
        for (int i = 1; i < probabilities.length; i++) {
            if (probabilities[i] > maxProb) {
                maxProb = probabilities[i];
                maxIndex = i;
            }
        }

        return new PredictionResult(
                ACTIVITY_LABELS[maxIndex],
                maxIndex,
                maxProb,
                probabilities
        );
    }

    /**
     * Compute softmax
     */
    private float[] softmax(float[] scores) {
        float[] probabilities = new float[scores.length];
        float maxScore = scores[0];

        // Find max for numerical stability
        for (float score : scores) {
            if (score > maxScore) maxScore = score;
        }

        // Compute exp and sum
        float sum = 0.0f;
        for (int i = 0; i < scores.length; i++) {
            probabilities[i] = (float) Math.exp(scores[i] - maxScore);
            sum += probabilities[i];
        }

        // Normalize
        for (int i = 0; i < probabilities.length; i++) {
            probabilities[i] /= sum;
        }

        return probabilities;
    }

    /**
     * Copy asset to cache directory
     */
    private String assetFilePath(Context context, String assetName) throws IOException {
        File file = new File(context.getFilesDir(), assetName);

        if (file.exists() && file.length() > 0) {
            return file.getAbsolutePath();
        }

        try (InputStream is = context.getAssets().open(assetName)) {
            try (OutputStream os = new FileOutputStream(file)) {
                byte[] buffer = new byte[4 * 1024];
                int read;
                while ((read = is.read(buffer)) != -1) {
                    os.write(buffer, 0, read);
                }
                os.flush();
            }
            return file.getAbsolutePath();
        }
    }

    public static class PredictionResult {
        public final String activityName;
        public final int activityIndex;
        public final float confidence;
        public final float[] allProbabilities;

        public PredictionResult(String activityName, int activityIndex,
                                float confidence, float[] allProbabilities) {
            this.activityName = activityName;
            this.activityIndex = activityIndex;
            this.confidence = confidence;
            this.allProbabilities = allProbabilities;
        }
    }

    public String[] getActivityLabels() {
        return ACTIVITY_LABELS;
    }
}

File: flutter_har_app/lib/models/har_model.dart
import 'dart:math';
import 'package:tflite_flutter/tflite_flutter.dart';
import 'package:flutter/services.dart';

class HARModel {
  Interpreter? _interpreter;
  static const int sequenceLength = 128;
  static const int numFeatures = 6;  // Only Accel(3) + Gyro(3)
  static const int numClasses = 6;

  // Activity labels matching UCI HAR dataset
  static const List<String> activityLabels = [
    "WALKING",
    "WALKING_UPSTAIRS",
    "WALKING_DOWNSTAIRS",
    "SITTING",
    "STANDING",
    "LAYING"
  ];

  // Static activity indices
  static const Set<int> staticActivityIndices = {3, 4, 5}; // SITTING, STANDING, LAYING

  // Temporal smoothing - keep history of predictions
  final List<PredictionResult> _predictionHistory = [];
  static const int maxHistorySize = 5;
  
  // Last sensor data for orientation analysis
  List<List<double>>? _lastSensorData;

  /// Initialize and load the TFLite model
  Future<void> loadModel() async {
    try {
      _interpreter = await Interpreter.fromAsset(
        'assets/models/har_model.tflite',
      );
      print('HAR Model loaded successfully');
    } catch (e) {
      print('Error loading model: $e');
      print('⚠️ Running in DEMO MODE with simulated predictions');
      // Don't throw - allow app to run in demo mode
    }
  }

  /// Predict activity from sensor data
  /// @param sensorData: List of shape [128, 6]
  /// @return PredictionResult containing activity and confidence
  Future<PredictionResult> predict(List<List<double>> sensorData) async {
    // Validate input
    if (sensorData.length != sequenceLength || sensorData[0].length != numFeatures) {
      throw ArgumentError(
        'Expected input shape [128, 6], got [${sensorData.length}, ${sensorData[0].length}]'
      );
    }

    // Store for orientation analysis
    _lastSensorData = sensorData;

    // If model not loaded, return simulated prediction
    if (_interpreter == null) {
      return _simulatePredict(sensorData);
    }

    try {
      // Prepare input tensor [1, 128, 6] - model has built-in normalization
      var input = List.generate(
        1,
        (_) => sensorData.map((row) => row.toList()).toList(),
      );

      // Prepare output tensor [1, 6] - properly initialized
      var output = List.generate(1, (_) => List<double>.filled(numClasses, 0.0));

      // Run inference
      _interpreter!.run(input, output);

      // Get probabilities - model already has softmax, direct use
      List<double> probabilities = List<double>.from(output[0]);
      
      // Debug: Log raw output from model
      double sum = probabilities.reduce((a, b) => a + b);
      print('HAR Model: Raw probabilities (sum=${sum.toStringAsFixed(4)}): '
            '${probabilities.map((p) => (p * 100).toStringAsFixed(1) + "%").join(", ")}');

      // Find max probability
      int maxIndex = 0;
      double maxProb = probabilities[0];
      
      for (int i = 1; i < probabilities.length; i++) {
        if (probabilities[i] > maxProb) {
          maxProb = probabilities[i];
          maxIndex = i;
        }
      }

      // Apply orientation-based refinement for static activities
      if (staticActivityIndices.contains(maxIndex)) {
        var orientationInfo = _analyzeOrientation(sensorData);
        maxIndex = _refineStaticActivity(maxIndex, probabilities, orientationInfo);
        maxProb = probabilities[maxIndex];
        print('HAR Model: Orientation-refined activity = ${activityLabels[maxIndex]}');
      }

      // Adjust confidence to be more realistic
      double adjustedConfidence = maxProb;
      if (maxProb > 0.95) {
        // Very confident predictions (>95%) get softened more
        adjustedConfidence = 0.70 + (maxProb - 0.95) * 5.0; // Maps 0.95-1.0 to 0.70-0.95
      } else if (maxProb > 0.85) {
        // Medium-high confidence stays roughly the same  
        adjustedConfidence = maxProb;
      }
      
      // Clamp to valid range
      adjustedConfidence = adjustedConfidence.clamp(0.0, 0.98);

      // Create result
      var result = PredictionResult(
        activityName: activityLabels[maxIndex],
        activityIndex: maxIndex,
        confidence: adjustedConfidence,
        allProbabilities: probabilities,
      );

      // Apply temporal smoothing for static activities
      result = _applyTemporalSmoothing(result);
      
      print('HAR Model: Final=${result.activityName} '
            'Confidence=${(result.confidence * 100).toStringAsFixed(1)}%');

      return result;
    } catch (e) {
      print('Error during prediction: $e');
      return _simulatePredict(sensorData);
    }
  }

  /// Analyze device orientation from accelerometer data
  Map<String, double> _analyzeOrientation(List<List<double>> sensorData) {
    // Calculate mean acceleration (gravity component)
    double sumX = 0, sumY = 0, sumZ = 0;
    for (var row in sensorData) {
      sumX += row[0];
      sumY += row[1];
      sumZ += row[2];
    }
    
    double meanX = sumX / sensorData.length;
    double meanY = sumY / sensorData.length;
    double meanZ = sumZ / sensorData.length;
    
    // Calculate magnitude and dominant axis
    double magnitude = sqrt(meanX * meanX + meanY * meanY + meanZ * meanZ);
    
    print('HAR Model: Orientation - X:${meanX.toStringAsFixed(2)}, '
          'Y:${meanY.toStringAsFixed(2)}, Z:${meanZ.toStringAsFixed(2)}, '
          'Mag:${magnitude.toStringAsFixed(2)}');
    
    return {
      'meanX': meanX,
      'meanY': meanY,
      'meanZ': meanZ,
      'magnitude': magnitude,
    };
  }

  /// Refine static activity prediction using orientation
  int _refineStaticActivity(int predictedIndex, List<double> probabilities, 
                            Map<String, double> orientation) {
    double meanX = orientation['meanX']!;
    double meanY = orientation['meanY']!;
    double meanZ = orientation['meanZ']!;
    
    // Find dominant axis (which has highest absolute value)
    double absX = meanX.abs();
    double absY = meanY.abs();
    double absZ = meanZ.abs();
    
    // LAYING: Device is horizontal (Z-axis dominant, X/Y has gravity)
    // SITTING/STANDING: Device is vertical (Z-axis has less gravity)
    
    // Simple heuristic:
    // If Z-axis dominates (device lying flat): likely LAYING
    // If X or Y axis dominates (device vertical): likely SITTING or STANDING
    
    double horizontalGravity = sqrt(absX * absX + absY * absY);
    double verticalGravity = absZ;
    
    bool isHorizontal = verticalGravity > horizontalGravity;
    
    // Get top 2 predictions from static activities
    double layingProb = probabilities[5];  // LAYING
    double sittingProb = probabilities[3]; // SITTING
    double standingProb = probabilities[4]; // STANDING
    
    int refinedIndex = predictedIndex;
    
    if (isHorizontal) {
      // Device horizontal: favor LAYING
      if (layingProb > 0.15) { // If LAYING has reasonable probability
        refinedIndex = 5; // LAYING
        print('HAR Model: Orientation suggests LAYING (horizontal)');
      }
    } else {
      // Device vertical: favor SITTING or STANDING
      if (sittingProb > standingProb && sittingProb > 0.15) {
        refinedIndex = 3; // SITTING
        print('HAR Model: Orientation suggests SITTING (vertical)');
      } else if (standingProb > 0.15) {
        refinedIndex = 4; // STANDING
        print('HAR Model: Orientation suggests STANDING (vertical)');
      }
    }
    
    return refinedIndex;
  }

  /// Apply temporal smoothing to reduce flickering predictions
  PredictionResult _applyTemporalSmoothing(PredictionResult currentResult) {
    // Add to history
    _predictionHistory.add(currentResult);
    if (_predictionHistory.length > maxHistorySize) {
      _predictionHistory.removeAt(0);
    }
    
    // If not enough history or dynamic activity, return as-is
    if (_predictionHistory.length < 3 || 
        !staticActivityIndices.contains(currentResult.activityIndex)) {
      return currentResult;
    }
    
    // For static activities, check consistency
    Map<int, int> activityCount = {};
    for (var pred in _predictionHistory) {
      activityCount[pred.activityIndex] = 
          (activityCount[pred.activityIndex] ?? 0) + 1;
    }
    
    // Find most common activity in history
    int mostCommonActivity = currentResult.activityIndex;
    int maxCount = 0;
    activityCount.forEach((activity, count) {
      if (count > maxCount) {
        maxCount = count;
        mostCommonActivity = activity;
      }
    });
    
    // If current prediction differs from trend and has low confidence, use trend
    if (mostCommonActivity != currentResult.activityIndex && 
        currentResult.confidence < 0.85 && 
        maxCount >= 2) {
      print('HAR Model: Smoothing applied - using ${activityLabels[mostCommonActivity]} '
            'based on history (occurred $maxCount/${_predictionHistory.length} times)');
      
      // Boost confidence slightly for consistent predictions
      double boostedConfidence = min(0.88, currentResult.confidence + 0.1);
      
      return PredictionResult(
        activityName: activityLabels[mostCommonActivity],
        activityIndex: mostCommonActivity,
        confidence: boostedConfidence,
        allProbabilities: currentResult.allProbabilities,
      );
    }
    
    return currentResult;
  }

  /// Simulate predictions based on sensor data patterns (DEMO MODE)
  PredictionResult _simulatePredict(List<List<double>> sensorData) {
    // Calculate movement intensity from accelerometer data
    double totalMovement = 0;
    for (var row in sensorData) {
      double ax = row[0];
      double ay = row[1];
      double az = row[2];
      totalMovement += (ax * ax + ay * ay + az * az).abs();
    }
    double avgMovement = totalMovement / sensorData.length;

    // Simulate activity detection based on movement
    int predictedActivity;
    if (avgMovement < 5) {
      predictedActivity = Random().nextBool() ? 3 : 4; // SITTING or STANDING
    } else if (avgMovement < 15) {
      predictedActivity = 0; // WALKING
    } else if (avgMovement < 25) {
      predictedActivity = Random().nextBool() ? 1 : 2; // UPSTAIRS or DOWNSTAIRS
    } else {
      predictedActivity = 5; // LAYING
    }

    // Generate realistic probabilities
    List<double> probabilities = List.filled(numClasses, 0.05);
    probabilities[predictedActivity] = 0.65 + Random().nextDouble() * 0.25;
    
    // Distribute remaining probability
    double remaining = 1.0 - probabilities[predictedActivity];
    for (int i = 0; i < numClasses; i++) {
      if (i != predictedActivity) {
        probabilities[i] = remaining / (numClasses - 1);
      }
    }

    return PredictionResult(
      activityName: activityLabels[predictedActivity],
      activityIndex: predictedActivity,
      confidence: probabilities[predictedActivity],
      allProbabilities: probabilities,
    );
  }

  // Normalization removed - model has built-in normalization layer

  /// Compute softmax
  List<double> _softmax(List<double> scores) {
    List<double> probabilities = List.filled(scores.length, 0.0);
    double maxScore = scores.reduce(max);

    // Compute exp and sum
    double sum = 0.0;
    for (int i = 0; i < scores.length; i++) {
      probabilities[i] = exp(scores[i] - maxScore);
      sum += probabilities[i];
    }

    // Normalize
    for (int i = 0; i < probabilities.length; i++) {
      probabilities[i] /= sum;
    }

    return probabilities;
  }

  /// Get activity labels
  List<String> getActivityLabels() {
    return activityLabels;
  }

  /// Dispose model resources
  void dispose() {
    _interpreter?.close();
    _interpreter = null;
    _predictionHistory.clear();
    _lastSensorData = null;
  }
}

class PredictionResult {
  final String activityName;
  final int activityIndex;
  final double confidence;
  final List<double> allProbabilities;

  PredictionResult({
    required this.activityName,
    required this.activityIndex,
    required this.confidence,
    required this.allProbabilities,
  });

  @override
  String toString() {
    return 'PredictionResult(activity: $activityName, confidence: ${(confidence * 100).toStringAsFixed(1)}%)';
  }
}
